README.md
# GrokApi

The is an experiment out of curiosity to see if Grok, X's Brand new AI can be used programmatically. Apparently it can!

Really cool use cases are:
- Grok as a voice-assistant to be used to control your home or vehicle, etc. 
- You can even ask it to output valid, well-formatted JSON enabling you to parse real-time events programmatically.

## Pre-requisites

You'll need to have a X premium+ membership. As of the time of writing this, Grok is in an early-access program, to which you can apply [here](https://grok.x.ai/).

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add grok_api

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install grok_api

## Usage
You'll need to set the following environment variables in the .env file in the root of your project:
 - `GROK_BEARER_TOKEN`. The `AAAAAAAAAAAAAAAAAAAA` etc. that comes after `Bearer`
 - `GROK_COOKIE`. Starts with `guest_id=v1%3A168`
 - `GROK_CSRF_TOKEN`. Started with `b4caa52c` in my case.

To get the values for these variables, go to X.com in the browser, access Grok (see pre-requisites) and open the Javascript Console. Make sure the Network panel is selected and the session is being recorded.
Then perform a query on grok. Filter on `Fetch/XHR` to find a valid request. Copy the request with headers. I used Copy as cURL.

You can apply this at the top of your script, as long as you don't share that section with others:
```ruby
ENV['GROK_BEARER_TOKEN'] = 'your_token_string_here'
ENV['GROK_COOKIE'] = 'your_cookie_string_here'
ENV['GROK_CSRF_TOKEN'] = 'your_csrf_token_here'
```

Then you can start a conversation and chat with Grok:
```ruby
client = GrokApi::Client.new
client.start_conversation
client.chat do
  say "Is the Improbability Drive incredible, or just highly improbable?"
  chat!
end
```

The chat will be printed by default:
```
2023-12-21 01:24:00 +0100: 💬 New conversation started on "Is the Improbability Drive incredible, or just hig"
2023-12-21 01:24:00 +0100: 💬 You: Is the Improbability Drive incredible, or just highly improbable?
2023-12-21 01:24:00 +0100: 🤖 Grok: The Improbability Drive is, in fact, incredibly improbable. It's so improbable that it's almost impossible, which is what makes it so incredible. It's like finding a functioning smartphone in a pile of rocks - highly improbable, yet quite incredible if it actually happens.
```

When you continue the chat by say more, it will continue the conversation with Grok:
```
2023-12-21 17:09:40 +0100: 💬 You: From what book was that reference, again?
2023-12-21 17:12:27 +0100: 🤖 Grok: The reference to the Improbability Drive is from the book "The Hitchhiker's Guide to the Galaxy" by Douglas Adams. It's a science fiction comedy series that follows the adventures of Arthur Dent, an unwitting human hitchhiker, and his alien friend, Ford Prefect, as they journey through the galaxy after Earth is destroyed to make way for an intergalactic bypass. The Improbability Drive is a key plot device in the series, allowing their spaceship, the Heart of Gold, to travel faster than light by manipulating probability.[link](#tweet=1736401424098165204)
```

```md
## ToDo's:
- [ ] Add a dot env example someone can copy (?).
- [ ] properly support a continued conversation (keep track of and include the conversation-stack)
- [ ] auto start a conversation if we don't have one yet `start_conversation unless @conversationId`
- [ ] use streaming API to parse messages during flushing of the output buffer
- [ ] add a chat interface `bin/chat`
- [ ] host an openAI compatible API?
- [ ] Proxy?

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/erikdebruijn/GrokApi. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/ErikDeBruijn/GrokApi/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the GrokApi project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/ErikDeBruijn/GrokApi/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.2] - 2023-12-22
- Able to have an extended chat instead of 1 request and 1 response.
- Environment variables can be set in a .env file.
- Added fun_mode option to the client. Enabled by default.
- Better specs in WebMock.

## [0.1.1] - 2023-12-21
- Grok's response will now be displayed and added to the conversation.
- Added a `say` method to the client.
- Added a `chat!` method to the client.

## [0.1.0] - 2023-12-20

- Initial release
  - It can get a `ConversationId` from X.
  - It can maintain a `Conversation`, which is currently just an array of messages.


spec/spec_helper.rb
# frozen_string_literal: true

require_relative "../lib/GrokAPI"
require_relative "../lib/GrokAPI/client"
require_relative "../lib/GrokAPI/conversation"

require "rspec"
require "vcr"
require "webmock/rspec"
require "dotenv/load"

VCR.configure do |c|
  c.cassette_library_dir = "spec/fixtures/vcr"
  c.hook_into :webmock
end


spec/GrokApi_spec.rb
# frozen_string_literal: true

RSpec.describe GrokApi do
  it "has a version number" do
    expect(GrokApi::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


spec/GrokApi/client_spec.rb
describe GrokApi::Client do
  let(:logger) { Logger.new($stdout) }

  describe "#start_conversation" do
    it "gets a conversationId" do
      stub_conversation_id
      client = GrokApi::Client.new(logger: logger)
      client.start_conversation
      expect(client.instance_variable_get(:@conversationId)).to eq("1234")
    end

    it "sends a message" do
      stub_conversation_id

      stub_request(:post, "https://api.twitter.com/2/grok/add_response.json")
        .with(
          body: "{\"responses\":[{\"message\":\"Is the Improbability Drive incredible, or just highly improbable?\",\"sender\":1}],\"systemPromptName\":\"\",\"conversationId\":\"1234\"}",
          headers: {
            "Accept" => "*/*",
            "Accept-Encoding" => "gzip;q=1.0,deflate;q=0.6,identity;q=0.3",
            "Content-Type" => "text/plain;charset=UTF-8",
            "Dnt" => "1",
            "Referer" => "https://twitter.com/",
            "Sec-Ch-Ua" => '"Not_A Brand";v="8", "Chromium";v="120"',
            "Sec-Ch-Ua-Mobile" => "?0",
            "User-Agent" => "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "X-Twitter-Client-Language" => "en"
          }
        )
        .to_return(
          status: 200,
          headers: {
            "Date" => "Thu, 21 Dec 2023 14:40:15 GMT",
            "Server" => "tsa_b",
            "Cache-Control" => "no-cache, no-store, max-age=0",
            "Strict-Transport-Security" => "max-age=631138519",
            "Transfer-Encoding" => "chunked"
          },
          body: <<~JSONLINES
            {"result":{"sender":"ASSISTANT","message":"The Impro","query":""}}
            {"result":{"sender":"ASSISTANT","message":"bability","query":""}}
            {"result":{"sender":"ASSISTANT","message":" Drive is","query":""}}
            {"result":{"sender":"ASSISTANT","message":" a fictional","query":""}}
            {"result":{"sender":"ASSISTANT","message":" concept from","query":""}}
            {"result":{"sender":"ASSISTANT","message":" Douglas Adams","query":""}}
            {"result":{"sender":"ASSISTANT","message":"'s","query":""}}
            {"result":{"sender":"ASSISTANT","message":" \\"The","query":""}}
            {"result":{"sender":"ASSISTANT","message":" Hitchh","query":""}}
            {"result":{"sender":"ASSISTANT","message":"iker'","query":""}}
            {"result":{"sender":"ASSISTANT","message":"s Guide","query":""}}
            {"result":{"sender":"ASSISTANT","message":" to the","query":""}}
            {"result":{"sender":"ASSISTANT","message":" Galaxy\\"","query":""}}
            {"result":{"sender":"ASSISTANT","message":" series.","query":""}}
            {"result":{"sender":"ASSISTANT","message":" In the","query":""}}
            {"result":{"sender":"ASSISTANT","message":" context of","query":""}}
            {"result":{"sender":"ASSISTANT","message":" the story","query":""}}
            {"result":{"sender":"ASSISTANT","message":", it","query":""}}
            {"result":{"sender":"ASSISTANT","message":" is an","query":""}}
            {"result":{"sender":"ASSISTANT","message":" incredible device","query":""}}
            {"result":{"sender":"ASSISTANT","message":" that allows","query":""}}
            {"result":{"sender":"ASSISTANT","message":" a spaceship","query":""}}
            {"result":{"sender":"ASSISTANT","message":" to pass","query":""}}
            {"result":{"sender":"ASSISTANT","message":" through every","query":""}}
            {"result":{"sender":"ASSISTANT","message":" point in","query":""}}
            {"result":{"sender":"ASSISTANT","message":" the Universe","query":""}}
            {"result":{"sender":"ASSISTANT","message":" simultaneously.","query":""}}
            {"result":{"sender":"ASSISTANT","message":" However,","query":""}}
            {"result":{"sender":"ASSISTANT","message":" as it","query":""}}
            {"result":{"sender":"ASSISTANT","message":" is a","query":""}}
            {"result":{"sender":"ASSISTANT","message":" fictional creation","query":""}}
            {"result":{"sender":"ASSISTANT","message":", it","query":""}}
            {"result":{"sender":"ASSISTANT","message":" is also","query":""}}
            {"result":{"sender":"ASSISTANT","message":" highly improbable","query":""}}
            {"result":{"sender":"ASSISTANT","message":" in the","query":""}}
            {"result":{"sender":"ASSISTANT","message":" real world","query":""}}
            {"result":{"sender":"ASSISTANT","message":".","query":""}}
            {"userChatItemId":"1737845452014309376","agentChatItemId":"1737845474835517440"}
          JSONLINES
      )

      client = GrokApi::Client.new(fun_mode: false, logger: Logger.new($stdout))
      client.start_conversation

      expect(client.instance_variable_get(:@conversationId).to_i).to be_positive

      client.chat do
        say "Is the Improbability Drive incredible, or just highly improbable?"
        chat!
      end

      expect(client.instance_variable_get(:@conversation).messages.last["message"]).to eq(<<~MARKDOWN.strip)
        The Improbability Drive is a fictional concept from Douglas Adams's "The Hitchhiker's Guide to the Galaxy" series. In the context of the story, it is an incredible device that allows a spaceship to pass through every point in the Universe simultaneously. However, as it is a fictional creation, it is also highly improbable in the real world.
      MARKDOWN
    end

    it "sends subsequent messages" do
      stub_conversation_id
      stub_request(:post, "https://api.twitter.com/2/grok/add_response.json")
        .with(
          body: '{"responses":[{"message":"Is the Improbability Drive incredible, or just highly improbable?","sender":1}],"systemPromptName":"fun","conversationId":"1234"}',
          headers: {
            "Accept" => "*/*",
            "Accept-Encoding" => "gzip;q=1.0,deflate;q=0.6,identity;q=0.3",
            "Content-Type" => "text/plain;charset=UTF-8",
            "Dnt" => "1",
            "Referer" => "https://twitter.com/",
            "Sec-Ch-Ua" => '"Not_A Brand";v="8", "Chromium";v="120"',
            "Sec-Ch-Ua-Mobile" => "?0",
            "User-Agent" => "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "X-Twitter-Client-Language" => "en"
          }
        )
        .to_return(
          status: 200,
          headers: {
            "Date" => "Thu, 21 Dec 2023 14:40:15 GMT",
            "Server" => "tsa_b",
            "Cache-Control" => "no-cache, no-store, max-age=0",
            "Strict-Transport-Security" => "max-age=631138519",
            "Transfer-Encoding" => "chunked"
          },
          body: <<~JSONLINES
            {"result":{"sender":"ASSISTANT","message":"The Impro","query":""}}
            {"result":{"sender":"ASSISTANT","message":"bability","query":""}}
            {"result":{"sender":"ASSISTANT","message":" Drive is","query":""}}
          JSONLINES
        )

      client = GrokApi::Client.new(logger: logger)
      client.start_conversation
      client.chat do
        say "Is the Improbability Drive incredible, or just highly improbable?"
        # VCR.use_cassette("conversation-1.chat1") do
          chat!
        # end
        say "From what book was that reference, again?"
        VCR.use_cassette("conversation-1.chat2") do
          chat!
        end
      end
    end
  end

  def stub_conversation_id
    stub_request(:get, "https://twitter.com/i/api/2/grok/conversation_id.json")
      .to_return(
        status: 200,
        headers: {
          "Set-Cookie" => "guest_id=v1%3A1234; Max-Age=63072000; Expires=Fri, 19 Dec 2025 22:10:41 GMT; Path=/; Domain=.twitter.com; Secure; SameSite=None",
          "Content-Type" => "application/json;charset=utf-8"
        },
        body: '{"conversationId":"1234"}'
      )
  end
end


lib/GrokApi/conversation.rb
module GrokApi
  class Conversation
    attr_reader :messages
    attr_accessor :topic

    def initialize(topic: nil)
      @messages = []
      @topic = nil
      @logger = Logger.new(STDOUT)
      @logger.formatter = proc do |severity, datetime, progname, msg|
        "#{datetime}: #{msg}\n"
      end
    end

    def <<(message)
      if @messages.empty?
        # trim message to first line, max. 50 characters:
        @topic = message.split("\n").first[0..49]
        @logger.info("💬 ____________________________________________________________")
        @logger.info("💬 New conversation started on \"#{@topic}\"")
      end
      @logger.info("💬 You: #{message}")
      @messages << { "message" => message, "sender" => 1 }
    end

    def assistant_response(message)
      @logger.info("🤖 Grok: #{message}")
      @messages << { "message" => message, "sender" => 2 }
    end
  end
end


lib/GrokApi/client.rb
# frozen_string_literal: true

require "faraday"

module GrokApi
  class Client
    def initialize(fun_mode: true, cookie: nil, bearer_token: nil, csrf_token: nil, logger: nil)
      @cookie = cookie || ENV["GROK_COOKIE"]
      @token = bearer_token || ENV["GROK_BEARER_TOKEN"]
      @csrf_token = csrf_token || ENV["GROK_CSRF_TOKEN"]
      @logger = logger || Logger.new($stderr)
      @logger.level = Logger::ERROR

      @connection = nil
      @conversationId = nil
      @fun_mode = fun_mode
      @conversation = Conversation.new
    end

    def start_conversation
      response = connection.get("/i/api/2/grok/conversation_id.json")
      raise "Response code #{response.status}" unless response.success?
      raise "Invalid response: #{response.body}" unless response.body["conversationId"]

      @conversationId = response.body["conversationId"]
      @conversation
    end

    def chat(&block)
      raise "No conversation started" unless @conversationId

      instance_eval(&block)
    end

    def say(message)
      @conversation << message
    end

    def chat!
      body = {
        "responses" => @conversation.messages,
        "systemPromptName" => @fun_mode ? "fun" : "",
        "conversationId" => @conversationId
      }.to_json

      response = Faraday.post("https://api.twitter.com/2/grok/add_response.json", body, api_headers)

      if response.success?
        # Parse the response body (which is in JSON Lines)
        responses = response.body.split("\n").map(&JSON.method(:parse))
        assistant_responses = responses.map do |line|
          line.dig("result", "message")
        end
        @conversation.assistant_response(assistant_responses.join(""))
      else
        # Handle the error
        @logger.error("Error: #{response.status}")
      end
    end

    private

    def api_headers
      {
        "sec-ch-ua" => '"Not_A Brand";v="8", "Chromium";v="120"',
        "dnt" => "1",
        "x-twitter-client-language" => "en",
        "x-csrf-token" => @csrf_token,
        "sec-ch-ua-mobile" => "?0",
        "authorization" => "Bearer #{@token}",
        "user-agent" => "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "content-type" => "text/plain;charset=UTF-8",
        "referer" => "https://twitter.com/",
        "cookie" => @cookie
      }
    end

    def connection
      @connection ||= Faraday.new(url: "https://twitter.com") do |faraday|
        faraday.headers["Host"] = "twitter.com"
        faraday.headers["Cookie"] = @cookie
        faraday.headers["Accept"] = "*/*"
        faraday.headers["Accept-Language"] = "en-US,en;q=0.9,nl;q=0.8"
        faraday.headers["Authorization"] = "Bearer #{@token}"
        faraday.headers["x-csrf-token"] = @csrf_token

        faraday.response :json
        # faraday.response :logger if @logger
      end
    end
  end
end
